MWI: Add `/wait` endpoint and `tbot wait` CLI helper

timothyb89 opened the pull request:
This adds two new features meant to improve use of `tbot` as a background service:
- A new `/wait(/service)` endpoint that functions similarly to `/readyz`, but waits to respond until the specified service (or all services) are initialized
- A new CLI helper, `tbot wait`, that requests this endpoint with a few robustness protections to make it safe and useful in bash scripts, like allowing it to start running before the diagnostics endpoint itself is ready and still return success appropriately.

Some small internal changes to tbot's readyz service were also made to track per-service initialization.

Note that as with `/readyz` the diagnostics service needs to be enabled via `--diag-addr` or similar for this functionality to be useful.

The HTTP API and the CLI helper have slightly different semantics for what "ready" means, as we want to minimize parsing and error handling for shell users:
- The HTTP API waits for the first readiness report from the service, after which it returns the latest status without delay. This means it may return when the service is unhealthy.
- The CLI helper waits for the service to actually report healthy. This means it may continue checking even once the wait endpoint returns a value.

Fixes #62117
changelog: Add `tbot wait` API and helper to let scripts wait for bots to become ready

---

### Usage examples

```code
# start a bot in the background
$ tbot start identity ... --diag-addr 127.0.0.1:4444

# wait for all services to become healthy
$ tbot wait --diag-addr 127.0.0.1:4444

# wait for a specific service to become healthy
$ tbot wait --diag-addr 127.0.0.1:4444 --service identity-1

# wait for a specific service to become healthy, with a timeout
$ tbot wait --diag-addr 127.0.0.1:4444 --service identity-1 --timeout 30s
```

timothyb89 committed:
MWI: Add `/wait` endpoint and `tbot wait` CLI helper

This adds two new features meant to improve use of `tbot` as a
background service:
- A new `/wait(/service)` endpoint that functions similarly to
  `/readyz/`, but waits to respond until specified service (or all
  services) are ready
- A new CLI helper, `tbot wait`, that requests this endpoint with a
  few protections to make it safe and useful in bash scripts.

Note that as with `/readyz` the diagnostics service needs to be
enabled via `--diag-addr` or similar for this functionality to be
useful.

The HTTP API and the CLI helper have slightly different semantics for
what "ready" means, as we want to minimize parsing and error handling
for shell users:
- The HTTP API waits for the first readiness report from the service,
  after which it returns the latest status without delay. This means
  it may return when the service is unhealthy.
- The CLI helper waits for the service to actually report healthy.
  This means it may continue checking even once the wait endpoint
  returns a value.

---

```code
$ tbot start identity ... --diag-addr 127.0.0.1:4444

$ tbot wait --diag-addr 127.0.0.1:4444

$ tbot wait --diag-addr 127.0.0.1:4444 --service identity-1

$ tbot wait --diag-addr 127.0.0.1:4444 --service identity-1 --timeout 30s
```

timothyb89 commented:
Outstanding items:

- [ ] Actually parse the returned status and make sure it reports healthy
- [ ] Tests

timothyb89 committed:
Parse and log detailed status info when available

timothyb89 committed:
Fix imports, remove unused error

timothyb89 committed:
Add tests and improve user-facing output from `tbot wait` helper

timothyb89 committed:
Fix imports

timothyb89 committed:
Remove unused test helpers

boxofrad commented on tool/tbot/wait_test.go:
```diff
@@ -0,0 +1,129 @@
+	"github.com/gravitational/teleport/lib/tbot/readyz"
+)
+
+func TestWaitTimeoutExceeded(t *testing.T) {
```
Given that we're testing the `readyz` HTTP handler without all of the other Teleport machinery, we could probably use `testing/synctest` effectively here to remove the possibility of any flakes caused by those `time.Sleep` calls.

Totally optional, though!

timothyb89 replied:
Good call, needed a custom roundtripper to avoid network IO but they run much faster now!

timothyb89 committed:
Switch to testing/synctest

This switches the tests depending on `time.Sleep()` to use
synctest along with an in-memory roundtripper.

timothyb89 commented on lib/tbot/readyz/readyz_test.go:
```diff
@@ -185,3 +192,236 @@ func TestAllServicesReported(t *testing.T) {
+				}()
+
+				// Sleep a bit so we actually wait
+				time.Sleep(time.Millisecond * 100)
```
Good idea, I forgot that package existed. I've swapped out the implementation, hopefully it's better now?

timothyb89 committed:
Fix error wording

timothyb89 committed:
Fix lints
